flatten_obj.py

```diff
@@ -0,0 +1,347 @@
+    cmd.flatten_obj(string name, string selection)
+
+SEE ALSO
+
+    split_states
+
+    """
+
+    # arguments
+
+    # Single argument; treat as selection
+    if name and not selection:
+        selection = name
+        name = ""
+    # default name and selection
+    if not name:
+        name = "flat"
+    if not selection:
+        selection = "(all)"
+
+    state = int(state)
+    rename = int(rename)
+    quiet = int(quiet)
+
+    # Wrap in extra parantheses for get_object_list
+    selection = "( %s )" % selection
+
+    if rename == 0:
+        chainSet = DefaultChainSet()
+    elif rename == 1:
+        chainSet = SequentialChainSet()
+    elif rename == 2:
+        chainSet = LongChainSet()
+    else:
+        raise ValueError("Unrecognized rename option (Valid: 0,1,2)")
+
+    metaprefix = "temp" #TODO unique prefix
+
+    # store original value of retain_order, which causes weird interleaving of
+    # structures if enabled.
+    retain_order = cmd.get("retain_order")
+    try:
+        cmd.set("retain_order",0)
+
+        # create new object for each state
+        for obj in cmd.get_object_list(selection):
+
+            if state <= 0:
+                # all states
+                prefix = "%s_%s_"%(metaprefix,obj)
+                cmd.split_states(obj,prefix=prefix)
+            else:
+                prefix = "%s_%s_%04d"%(metaprefix,obj,state)
+                cmd.create(prefix, obj, state, 1)
+
+        # renumber all states
+        statere = re.compile("^%s_(.*)_(\d+)$" % metaprefix) # matches split object names
+
+        warn_lowercase = False
+
+        # Iterate over all objects with metaprefix
+        try:
+            for obj in cmd.get_object_list("(%s_*)"%(metaprefix) ):
+                m = statere.match(obj)
+                if m is None:
+                    print("Failed to match object %s" %obj)
+                    continue
+                origobj = m.group(1)
+                statenum = int(m.group(2))
+
+                chains = set([atom.chain for atom in cmd.get_model(obj).atom])
+
+                rev_chain_map = {} #old -> new, for this obj only
+                for chain in sorted(chains):
+                    new_chain = chainSet.map_chain(origobj,statenum,chain)
+                    rev_chain_map[chain] = new_chain
+                    if not quiet:
+                        print("  %s state %d chain %s -> %s"%(origobj,statenum,chain, new_chain) )
+                    if not _long_chains:
+                        if len(new_chain) > 1:
+                            raise OutOfChainsError("No additional chains available (max 62).")
+
+                try:
+                    stored_name = stored.get_unused_name()
+                except AttributeError:
+                    # backup solution for PyMOL < 1.6
+                    stored_name = "_flatten_obj_chainMapping"
+
+                setattr(stored,stored_name,rev_chain_map)
+                cmd.alter(obj,"chain = stored.%s[chain]"%stored_name)
```

Comment:
using the "space" argument with alter (or iterate) is a more elegant solution than the global "stored" object. See http://pymolwiki.org/index.php/Iterate#PYMOL_API